04.class/cli.js

```diff
@@ -0,0 +1,121 @@
+import minimist from "minimist";
+import readline from "readline";
+import Memo from "./memo.js";
+import MemoHandler from "./memoHandler.js";
+
+const argv = minimist(process.argv.slice(2));
+
+async function main() {
+  const memo = new Memo();
+  const memoHandler = new MemoHandler();
+
+  if (!process.stdin.isTTY || Object.keys(argv).length < 2) {
+    try {
+      const lines = await readStdin();
+      const title = await lines[0];
+      const content = await lines.join("\n");
+
+      await memo.createTable();
+      await memo.insert(title, content);
+      console.log(`✅ タイトル "${title}" のメモが追加されました。`);
+    } catch (err) {
+      if (err instanceof Error) {
+        console.log("メモを追加できませんでした。error:", err.message);
+      } else {
+        throw err;
+      }
+    }
+  } else if (argv.l) {
+    try {
+      const memos = await memo.fetchAll();
+
+      if (memos.length > 0) {
+        memos.forEach((memo) => {
+          console.log(memo.title);
+        });
+      } else {
+        console.log("メモがありません");
+      }
+    } catch (err) {
+      if (err instanceof Error) {
+        console.log("メモの一覧の取得に失敗しました。error:", err.message);
+      } else {
+        throw err;
+      }
+    }
+  } else if (argv.r) {
+    try {
+      const memos = await memo.fetchAll();
+
+      if (memos.length > 0) {
+        const answer = await memoHandler.pickup(
+          memos,
+          "Choose a note you want to see:",
+        );
+        const selectedMemoData = await memo.select(answer.id);
+        console.log(selectedMemoData.content);
+      } else {
+        console.log("メモがありません");
+      }
+    } catch (err) {
+      if (err instanceof Error) {
+        console.log("メモの取得に失敗しました。error:", err.message);
+      } else {
+        throw err;
+      }
+    }
+  } else if (argv.d) {
+    try {
+      const memos = await memo.fetchAll();
+
+      if (memos.length > 0) {
+        const answer = await memoHandler.pickup(
+          memos,
+          "Choose a note you want to delete:",
+        );
+        await memo.delete(answer.id);
+        console.log(`🗑️  タイトル "${answer.title}" のメモが削除されました。`);
+      } else {
+        console.log("メモがありません");
+      }
+    } catch (err) {
+      if (err instanceof Error) {
+        console.log("メモの削除に失敗しました。error:", err.message);
+      } else {
+        throw err;
+      }
+    }
+  }
+}
+
+function readStdin() {
+  return new Promise((resolve, reject) => {
+    const lines = [];
+    let reader;
+
+    if (!process.stdin.isTTY || Object.keys(argv).length < 2) {
+      reader = readline.createInterface({
+        input: process.stdin,
+        output: process.stdout,
+      });
+    } else {
+      reader = readline.createInterface({
+        input: process.stdin,
+      });
+    }
+
+    reader.on("line", (line) => {
+      lines.push(line);
+    });
+
+    reader.on("close", () => {
+      if (!lines.every((line) => line.trim() === "")) {
```

Comment:
否定の位置を調整したほうが意味が分かりやすくなりそうです。元の条件式だと every の中の条件式がどうなるかと every の結果がどうなるか、それを反転するとどうなるかを考える必要がありますが、新しい条件式だと every の結果を反転する必要がなくなるからです。

```suggestion
      if (lines.every((line) => line.trim() !== "")) {
```

Reply:
新しい式の場合`lines`(配列)のすべての要素が空文字以外の場合、false となるため空行を入れたメモが追加できなくなるかと思います。元の条件式では、`lines`(配列)のすべての要素が空白文字でない限り、true になるのでこのままでも良いかと思いました

Reply:
すみません、`every` を `some` に修正するのを忘れていました。

```suggestion
      if (lines.some((line) => line.trim() !== "")) {
```

空文字であるかをチェックしたいのではなく、文字が入力されていることをチェックしたいのだと思うので、以下のように書いたほうがより分かりやすくなるかもしれません。

```suggestion
      if (lines.some((line) => line.trim().length > 0)) {
```

Reply:
someメソッドを使うようにしました